keps/168-2-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,292 @@
+
+Similarly to the ```metrics-server``` the server will be implemented with the [apiserver library](https://github.com/kubernetes/apiserver), which provides authentication and authorization.
+
+### API Details
+
+We introduce a new API that will extend the existing one.
+
+There will be separate endpoints for administrators and regular users. Each endpoint exposes information about a pending workload, such as:
+- workload's position in a ClusterQueue,
+- workload's position in a LocalQueue,
+- workload's priority,
+- creation timestamp.
+
+The API does not allow for the modification of any objects.
+
+Regular users will have access only to the LocalQueues they are assigned to. However, they will be able to fetch information about the global position of a workload in a ClusterQueue, without any details about workloads in different LocalQueues. 
+
+Administrators will have access to all the data at the ClusterQueue level. They will be able to view all the workloads, no matter the LocalQuees the workloads are assigned to. 
+
+### URIs:
+
+We introduce a new API group ```pending-workloads.kueue.x-k8s.io``` that aggregates the following endpoints:
+
+#### List all pending workloads in ClusterQueue
+
+```
+GET /apis/pending-workloads.kueue.x-k8s.io/VERSION/clusterqueues/CQ_NAME
+```
+
+#### List all pending workloads in LocalQueue
+```
+GET /apis/pending-workloads.kueue.x-k8s.io/VERSION/namespaces/LQ_NAMESPACE/localqueues/LQ_NAME
+```
+
+#### Fetch information about single Workload
+```
+GET /apis/pending-workloads.kueue.x-k8s.io/VERSION/namespaces/WL_NAMESPACE/workloads/WL_NAME
+```
+
+### API Objects:
+
+```
+// User-facing representation of a PendingWorkload in both LocalQueues and ClusterQueue that summarizes neccessary information from the admission order perspective
+type PendingWorkloadSummary struct {
+	TypeMeta TypeMeta
+	ObjectMeta ObjectMeta
+
+	LocalQueueNamestring
+	PositionInClusterQueue int32
+	PositionInLocalQueue int32
+	Priority int32
+}
+
+type PendingWorkloadsSummaryList struct {
+	TypeMeta TypeMeta
+	ListMeta ListMeta
+
+	Items []PendingWorkloadSummaryInClusterQueue
+}
+```
+
+### KueueManager changes
+
+We propose changing KueueManager's cache, so it contains more information about pending workloads. This allows us to expose that information. We'll need cache:
```

Comment:
Maybe this is not needed at this point. For example instead of using the cache structure the queue manager could implement `ComputeQueueHead(queueName, maxCount) []*workload.Info`, bypassing the cache. Still, it might be desirable to reuse the cache, but maybe we can defer this for later, based on user / review feedback.